src/timestamp9.c

```diff
@@ -226,33 +226,41 @@ timestamp9_in(PG_FUNCTION_ARGS)
 		time_t tt;
 
 		num_read = sscanf(str, "%d-%d-%d %d:%d:%d.%lld %c%5s", &tm_.tm_year, &tm_.tm_mon, &tm_.tm_mday, &tm_.tm_hour, &tm_.tm_min, &tm_.tm_sec, &ns, &plusmin, gmt_offset_str);
-		if (num_read == 9 && fractional_valid)
+		if ((num_read == 7 || num_read == 9) && fractional_valid)
 		{
-			bool offset_valid = false;
-			gmt_offset = parse_gmt_offset(gmt_offset_str, &offset_valid);
-
-			if (!offset_valid)
+			if (num_read == 9)
 			{
-				ereport(ERROR,
+				/* If we specify the timezone, try to decode it. */
+				if (DecodeTimezone(psprintf("%c%s", plusmin, gmt_offset_str), &gmt_offset) != 0)
+				{
+					ereport(ERROR,
 						(errcode(ERRCODE_DATETIME_VALUE_OUT_OF_RANGE),
-							errmsg("invalid input format for timestamp9: could not parse gmt offset, required format y-m-d h:m:s.ns +tz \"%s\"",
-								   str)));
+						 errmsg("invalid input format for timestamp9: could not parse gmt offset, required format y-m-d h:m:s.ns [+tz] \"%s\"",
+							str)));
+				}
+			}
+			else if (num_read == 7)
```

Comment:
how does this handle bad input? eg. `select '2022-12-30 13:00:00.123456789 abc'::timestamp9;`

I wonder if the approach taken here in decoding shouldn't be a bit different. It could start by decoding the `%d-%d-%d %d:%d:%d.%lld` part. Then, if there's a remaining string, it could try `DecodeTimezone` on that part. You could use `%n` in the sscanf string to get an offset into the original string (this can avoid your psprintf call).

In that approach, I think we should then also support `DecodeTimezoneAbbrev` (try numeric decode first, if it fails we can still try abbrev?)

Reply:
Unfortunately, `DecodeTimezone()` cannot handle the below malformed input, it will treat the single plus sign or minus sign as '+0000' or '-0000'. The builtin timestamp data type will reject it. We need to reject it by ourselves.

`select '2023-01-25 00:00:00.123456789 +'::timestamp9`,
`select '2023-01-25 00:00:00.123456789 -'::timestamp9`

Reply:
It seems that the function `parse_sane_timezone` from `backend/utils/adt/timestamp.c` is very handy to parse time zone information. But it's a static function.